windows/src/engine/kmtip/kmkey.cpp

```diff
@@ -243,10 +264,51 @@ HRESULT WINAPI CKeymanEditSession::KeymanGetContext(int n, PWSTR buf)   // I3567
     return S_FALSE;
   }
 
-  if (!GetLeftOfSelection(_ec, _pContext, buf, n)) {   // I4933
+  if (!GetLeftOfSelection(_ec, _pContext, buf, n, &hrGetSelection, &cFetched, &tfSelection)) {  // I4933
     return S_FALSE;   // I4933
   }
 
+  // now check for selected text
+  if (!SUCCEEDED(hr = KeymanIsTextSelected(hrGetSelection, cFetched, tfSelection, isTextSelected))) {
+    Log(L"KeymanGetContext: Warning: KeymanIsTextSelected failed");
+    // Continue to return S_OK, even though checking if a text selection exists has failed.
+    // Reaching this point means the context in 'buf' is valid and will be passed to the caller.
+    // isTextSelected will be false, which is better than returning E_FAIL at this point.
+  }
+  return S_OK;
+}
+
+HRESULT
+CKeymanEditSession::KeymanIsTextSelected(HRESULT hrGetSelection, ULONG cFetched, TF_SELECTION tfSelection, BOOL *isTextSelected) {
```

Comment:
It feels weird to be passing hrGetSelection and cFetched in to this function. It seems like we could avoid a bunch of complexity by making this a simple BOOL return value if we didn't have to control for errors -- you don't use the hr result in the caller, so just log failure within the function and return TRUE/FALSE.


```c
  // now check for selected text
  isTextSelected = KeymanIsTextSelected(hrGetSelection, cFetched, tfSelection);

  if(tfSelection.range) {
    tfSelection.range->Release();
  }

  // ...

BOOL
CKeymanEditSession::KeymanIsTextSelected(HRESULT hrGetSelection, ULONG cFetched, TF_SELECTION tfSelection) {
  if (hrGetSelection == TF_E_NOSELECTION || cFetched == 0 || !SUCCEEDED(hrGetSelection) || !tfSelection.range) {
    // No selection is present, or an error occurred trying to get the selection
    return FALSE;
  }

  // check if the range is empty
  HRESULT hr;
  BOOL isEmpty = TRUE;

  // Compare the start and end of the range
  if (!SUCCEEDED(hr = tfSelection.range->IsEmpty(_ec, &isEmpty))) {
    SendDebugMessageFormat(L"KeymanIsTextSelected: Exit: Testing range->isEmpty failed with %x", hr);
    return FALSE;
  }

  return !isEmpty;
}
```